specs/gloas/beacon-chain.md

```diff
@@ -326,38 +394,50 @@ class BeaconState(Container):
 
 ### Predicates
 
-#### New `is_builder_withdrawal_credential`
+#### New `is_builder`
 
 ```python
-def is_builder_withdrawal_credential(withdrawal_credentials: Bytes32) -> bool:
-    return withdrawal_credentials[:1] == BUILDER_WITHDRAWAL_PREFIX
+def is_builder(state: BeaconState, pubkey: BLSPubkey) -> bool:
+    builder_pubkeys = [v.pubkey for v in state.builders]
+    return pubkey in builder_pubkeys
 ```
 
-#### New `has_builder_withdrawal_credential`
+#### New `is_validator`
 
 ```python
-def has_builder_withdrawal_credential(validator: Validator) -> bool:
-    """
-    Check if ``validator`` has an 0x03 prefixed "builder" withdrawal credential.
-    """
-    return is_builder_withdrawal_credential(validator.withdrawal_credentials)
+def is_validator(state: BeaconState, pubkey: BLSPubkey) -> bool:
+    validator_pubkeys = [v.pubkey for v in state.validators]
+    return pubkey in validator_pubkeys
 ```
 
-#### Modified `has_compounding_withdrawal_credential`
+#### New `is_builder_index`
 
-*Note*: The function `has_compounding_withdrawal_credential` is modified to
-return true for builders.
+```python
+def is_builder_index(validator_index: ValidatorIndex) -> bool:
+    return (validator_index & BUILDER_INDEX_FLAG) != 0
+```
+
+#### New `is_active_builder`
 
 ```python
-def has_compounding_withdrawal_credential(validator: Validator) -> bool:
+def is_active_builder(state: BeaconState, builder_index: BuilderIndex) -> bool:
     """
-    Check if ``validator`` has an 0x02 or 0x03 prefixed withdrawal credential.
+    Check if the builder at ``builder_index`` is active for the given ``state``.
     """
-    if is_compounding_withdrawal_credential(validator.withdrawal_credentials):
-        return True
-    if is_builder_withdrawal_credential(validator.withdrawal_credentials):
-        return True
-    return False
+    builder = state.builders[builder_index]
+    return (
+        # Placement in builder list is finalized
+        builder.deposit_epoch <= state.finalized_checkpoint.epoch
```

Comment:
This check misses the case when epoch is finalized while the deposit is actually not yet. Potentially can switch to either
```suggestion
        builder.deposit_slot <= compute_start_slot_at_epoch(state.finalized_checkpoint.epoch)
```
or
```suggestion
        builder.deposit_epoch < state.finalized_checkpoint.epoch
```